---
output: html_fragment
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Motivation

Proteomics data is collected on instruments from multiple vendors with various acquisition methods. Similiarly, analysis of the raw LC-MS/MS data is carried out using a variety of software packages, including both commercial and open-source tools. Each of these tools implements steps of the statistical analysis, including normalization (both at the peptide and/or protein level) and differential expression analyses. However, each package accomplishes these tasks in slightly different ways, and often limited approaches for normalization and differential expression are provided. These differences also make it more difficult to directly compare the effects of analyzing a dataset with different software packages. Lastly, improvements in instrument throughput and sensitivity generate large datasets that require computational skills and tools to effectively explore and interpret experimental results.

ProteoVista is a secondary analysis pipeline that combines two open-source tools to provide a standard workflow to process data from multiple primary analysis tools and provide an interactive dashboard for exploratory data analysis. The first part of the pipeline uses the R package [__MS-DAP__](https://github.com/ftwkoopmans/msdap) published by [Koopmans et al, J. Proteome Res 2023](https://pubs.acs.org/doi/10.1021/acs.jproteome.2c00513). As outlined in the figure below, MS-DAP provides tools for processing both data dependent acquisition (DDA) and data independent acquisition (DIA) dataset that have been searched using various tools including [Spectronaut (DIA)](https://biognosys.com/software/spectronaut/), [FragPipe/IonQuant (DDA) and FragPipe/DIA-NN (DIA)](https://fragpipe.nesvilab.org), [ProteomeDiscoverer (DDA and DIA)](https://www.thermofisher.com/order/catalog/product/B51001473) and others. After importing peptide-level information, MS-DAP provides multiple strategies for normalization and differential expression testing.

![MS-DAP overview.](images/msdap-overview.png)

<br/><br/>

After generating the protein-level quantitation and differential expression testing with MS-DAP, these results are then formatted for upload into the [__Quickomics__](https://quickomics.bxgenomics.com) Shiny dashboard for interactive data exploration. This tool was developed to provide a single interface to explore the results of RNA-seq and proteomics datasets, including various standard QC plots including PCA, volcano plots, heatmaps, expression box/violin plots.
